newlib/libc/sys/vita/select.c

```diff
@@ -0,0 +1,128 @@
+/*
+
+Copyright (C) 2016, Sunguk Lee
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+
+*/
+#include <sys/select.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <psp2/kernel/threadmgr.h>
+#include <psp2/net/net.h>
+#include <psp2/types.h>
+
+#include "vitadescriptor.h"
+
+#define SCE_ERRNO_MASK 0xFF
+
+#define MAX_EVENTS 255
+
+#define REG_POLL(_fd, event) \
+	do {														\
+		SceNetEpollEvent ev = {0};								\
+		ev.data.fd = (_fd);										\
+		ev.events = (event);									\
+		DescriptorTranslation *fdmap = __vita_fd_grab((_fd));	\
+		if (!fdmap) {											\
+			break;												\
+		}														\
+		sceNetEpollControl(eid, SCE_NET_EPOLL_CTL_ADD, fdmap->sce_uid, &ev);	\
+		__vita_fd_drop(fdmap);									\
+	} while(0)
+
+int select(int nfds, fd_set *readfds, fd_set *writefds, fd_set *exceptfds,
```

Comment:
this function should be thread-safe. not sure if it's worth adding thread-safety here though, ideally we'd grab all required FDs before creating the epoll, and drop them after wait returns.